src/neo/Network/P2P/Payloads/TransactionAttributeCollection.cs

```diff
@@ -0,0 +1,134 @@
+    {
+        /// <summary>
+        /// Maximum number of attributes that can be contained within a transaction
+        /// </summary>
+        public const int MaxTransactionAttributes = 16;
+
+        private readonly Dictionary<TransactionAttributeType, List<TransactionAttribute>> _entries;
+
+        public int Size =>
+            IO.Helper.GetVarSize(_entries.Count) +      // count
+            _entries.Values.Sum(u => u.GetVarSize());   // entries
+
+        public int Count => _entries.Count;
+
+        public Cosigner[] Cosigners
+        {
+            get
+            {
+                if (_entries.TryGetValue(TransactionAttributeType.Cosigner, out var cosigners))
+                {
+                    return cosigners.Cast<Cosigner>().ToArray();
+                }
+
+                return Array.Empty<Cosigner>();
+            }
+        }
+
+        public TransactionAttributeCollection()
+        {
+            _entries = new Dictionary<TransactionAttributeType, List<TransactionAttribute>>();
+        }
+
+        public TransactionAttributeCollection(params TransactionAttribute[] attributes)
+        {
+            _entries = new Dictionary<TransactionAttributeType, List<TransactionAttribute>>();
+
+            if (attributes != null)
+            {
+                foreach (var attr in attributes)
+                {
+                    Add(attr);
+                }
+            }
+        }
+
+        public bool TryGet<T>(TransactionAttributeType type, out T[] attr)
+        {
+            if (_entries.TryGetValue(type, out var val))
+            {
+                attr = val.Cast<T>().ToArray();
+                return true;
+            }
+
+            attr = null;
+            return false;
+        }
+
+        public void Serialize(BinaryWriter writer)
+        {
+            writer.WriteVarInt(_entries.Sum(u => u.Value.Count));
+            foreach (var attr in _entries)
+            {
+                foreach (var entry in attr.Value)
+                {
+                    writer.Write(entry);
+                }
+            }
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            _entries.Clear();
+
+            var count = (int)reader.ReadVarInt(MaxTransactionAttributes);
+            for (int x = 0; x < count; x++)
+            {
+                Add(TransactionAttribute.DeserializeFrom(reader));
+            }
+
+            // Check duplicate cosigners
+
+            var cosigners = Cosigners;
+            if (cosigners.Select(u => u.Account).Distinct().Count() != cosigners.Length) throw new FormatException();
+        }
+
+        public void Add(TransactionAttribute attr)
+        {
+            if (_entries.TryGetValue(attr.Type, out var list))
+            {
+                list.Add(attr);
+            }
+            else
+            {
+                _entries[attr.Type] = new List<TransactionAttribute>(new TransactionAttribute[] { attr });
+            }
+        }
```

Comment:
After serialization and deserialization, the order of attributes may change, resulting in a change in the hash of the transaction.

Reply:
Yes, you are right, we should use a list